app/models/source_resolver.rb

```diff
@@ -1,34 +1,62 @@
 class SourceResolver
-  def self.resolve(new_source)
-    new(new_source).resolve
+  def self.resolve(adjusted_source)
+    new(adjusted_source).resolve
   end
 
-  def initialize(new_source)
-    @new_source = new_source
+  def initialize(adjusted_source)
+    @adjusted_source = adjusted_source
   end
 
   def resolve
-    return if @new_source.save
+    return if @adjusted_source.save
     adjust_ranks
-    @new_source.save
+    @adjusted_source.save
   end
 
   private
 
+  def conflicted_rank_types
+    @adjusted_source.errors.keys
+  end
+
   def adjust_ranks
     conflicted_rank_types.each do |rank_type|
-      sources_to_adjust(rank_type).each do |source|
-        source.increment!(rank_type)
+      if demotion?(rank_type)
+        move_sources_up(rank_type)
+      else
+        move_sources_down(rank_type)
       end
     end
   end
 
-  def conflicted_rank_types
-    @new_source.errors.keys
+  def demotion?(rank_type)
+    return false unless @adjusted_source.persisted?
+
+    old, new = @adjusted_source.changes[rank_type]
+    old < new
+  end
+
+  def move_sources_up(rank_type)
+    sources_to_decrement(rank_type).each do |source|
+      source.decrement!(rank_type)
+    end
+  end
+
+  def sources_to_decrement(rank_type)
+    cutoff = @adjusted_source.rank_for rank_type
+    Source.where("#{rank_type} <= ?", cutoff).order(rank_type)
+  end
+
+  def move_sources_down(rank_type)
+    sources_to_increment(rank_type).each do |source|
+      source.increment!(rank_type)
+    end
   end
 
-  def sources_to_adjust(rank_type)
-    cutoff = @new_source.rank_for rank_type
-    Source.where("#{rank_type} >= ?", cutoff).order("#{rank_type} DESC")
+  def sources_to_increment(rank_type)
+    cutoff = @adjusted_source.rank_for rank_type
+    old_rank = @adjusted_source.changes[rank_type].first || Source.count + 1
```

Comment:
I'm curious about this `||` switch. In what case does `adjusted_source.changes[rank_type]` come up `nil`?